raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -1487,39 +1510,57 @@ public void poll() throws IOException {
     }
 
     private void handlePendingAppends(long currentTimeMs) {
-        UnwrittenAppend unwrittenAppend = unwrittenAppends.poll();
-        if (unwrittenAppend == null || unwrittenAppend.isCancelled())
-            return;
-
         if (quorum.isLeader()) {
-            if (unwrittenAppend.isTimedOut(currentTimeMs)) {
-                unwrittenAppend.fail(new TimeoutException());
-            } else {
-                LeaderState leaderState = quorum.leaderStateOrThrow();
-                int epoch = quorum.epoch();
-                LogAppendInfo info = maybeAppendAsLeader(leaderState, unwrittenAppend.records);
-                if (info != null) {
-                    OffsetAndEpoch offsetAndEpoch = new OffsetAndEpoch(info.lastOffset, epoch);
-                    unwrittenAppend.complete(offsetAndEpoch);
-                    uncommittedAppends.put(offsetAndEpoch, new AppendBatchAndTime(info.lastOffset - info.firstOffset + 1, currentTimeMs));
+            for (UnwrittenAppend unwrittenAppend: unwrittenAppends) {
+                if (unwrittenAppend.isCancelled())
+                    continue;
+
+                if (unwrittenAppend.isTimedOut(currentTimeMs)) {
+                    unwrittenAppend.fail(new TimeoutException());
                 } else {
-                    unwrittenAppend.fail(new InvalidRequestException("Leader refused the append"));
+                    LeaderState leaderState = quorum.leaderStateOrThrow();
+                    int epoch = quorum.epoch();
+                    LogAppendInfo info = maybeAppendAsLeader(leaderState, unwrittenAppend.records);
+                    if (info != null) {
+                        OffsetAndEpoch offsetAndEpoch = new OffsetAndEpoch(info.lastOffset, epoch);
+
+                        if (unwrittenAppend.commitMode == AppendCommitMode.LEADER) {
+                            unwrittenAppend.complete(offsetAndEpoch);
+                        } else {
+                            appendPurgatory.await(unwrittenAppend.future, unwrittenAppend.requestTimeoutMs);
```

Comment:
There's a small catch that we do not count int the time from `append` call to the actual appending operation, i.e. the time when it was buffered in the unwritten queue, but I think this is not a very big issue.